gnomad/utils/transcript_annotation.py

```diff
@@ -71,3 +71,77 @@ def summarize_rsem_mt(
         )
 
     return transcript_ht.key_by("transcript_id", "gene_id"), gene_ht.key_by("gene_id")
+
+
+def get_expression_proportion(
+    transcript_ht: hl.Table,
+    gene_ht: hl.Table,
+    tissues_to_filter: Optional[List[str]] = None,
+    tissue_as_row: bool = False,
+) -> hl.Table:
+    """
+    Calculate the proportion of expression of transcript to gene per tissue.
+
+    :param transcript_ht: Table of summarized transcript expression by tissue
+    :param gene_ht: Table of summarized gene expression by tissue
+    :param tissues_to_filter: Optional list of tissues to filter out
+    :param tissue_as_row: If True, the input Table is with a row annotation for each
+        tissue instead of an array of values. Default is False.
+    :return: Table with expression proportion of transcript to gene per tissue
+        and mean expression proportion across tissues
+    """
+    if tissue_as_row:
+        tissues1 = list(gene_ht.row)
+        tissues1.remove("gene_id")
+        gene_ht = gene_ht.select(
+            gene_expression=hl.array([gene_ht[tissue] for tissue in tissues1])
+        ).annotate_globals(tissues=tissues1)
+
+        tissues2 = list(transcript_ht.row)
+        tissues2.remove("transcript_id")
+        tissues2.remove("gene_id")
+        transcript_ht = transcript_ht.select(
+            transcript_expression=hl.array(
+                [transcript_ht[tissue] for tissue in tissues2]
+            )
+        ).annotate_globals(tissues=tissues2)
+
+    # Join the transcript expression table and gene expression table
+    transcript_ht = transcript_ht.annotate(
+        gene_expression=gene_ht[transcript_ht.gene_id].gene_expression
+    )
+
+    if tissues_to_filter is not None:
+        tissues_indices = []
+        for i, t in enumerate(hl.eval(transcript_ht.tissues)):
+            if t not in tissues_to_filter:
+                tissues_indices.append(i)
+
+        transcript_ht = transcript_ht.annotate(
+            transcript_expression=hl.array(
+                [transcript_ht.transcript_expression[i] for i in tissues_indices]
+            ),
+            gene_expression=hl.array(
+                [transcript_ht.gene_expression[i] for i in tissues_indices]
+            ),
+        )
+
+        transcript_ht = transcript_ht.select_globals(
+            tissues=hl.array([transcript_ht.tissues[i] for i in tissues_indices])
+        )
+
+    # Calculate the proportion of expression of transcript to gene per tissue
+    transcript_ht = transcript_ht.annotate(
+        exp_prop=hl.or_else(
+            transcript_ht.transcript_expression / transcript_ht.gene_expression,
```

Comment:
Ah, I see now, I had in my mind that `summarize_transcript_expression` was running `summary_agg_func` on the grouped `gene_id` rather than `hl.agg.sum`.  Now that I look at this I think we probably want to change the other function because that is confusing. 


There are two ways we can do this. 

The first is to change the other function so it returns a single table with structs for `transcript_expression` and `expression_proportion`. Then modify the other functions accordingly. Something like this:

```
def summarize_transcript_expression(
    mt: hl.MatrixTable,
    transcript_expression_expr: Union[hl.expr.NumericExpression, str] = "transcript_tpm",
    tissue_expr: Union[hl.expr.StringExpression, str] = "tissue",
    summary_agg_func: Optional[Callable] = None,
) -> hl.Table:
    if summary_agg_func is None:
        summary_agg_func = lambda x: hl.median(hl.agg.collect(x))

    if isinstance(transcript_expression_expr, str):
        transcript_expression_expr = mt[transcript_expression_expr]

    if isinstance(tissue_expr, str):
        tissue_expr = mt[tissue_expr]

    mt = mt.group_cols_by(tissue=tissue_expr).aggregate(
        tx=summary_agg_func(transcript_expression_expr)
    )

    transcript_ht = mt.rename({"tx": ""}).make_table()
    transcript_ht = transcript_ht.key_by("transcript_id", "gene_id")

    tissues = list(transcript_ht.row_value)
    
    # Calculate the sum of transcript expression by gene per tissue.
    gene_ht = transcript_ht.group_by("gene_id").aggregate(
        **{
            tissue: hl.agg.sum(transcript_ht[tissue])
            for tissue in tissues
        }
    )

    # Calculate the proportion of expression of transcript to gene per tissue.
    gene_prop = gene_ht[transcript_ht.gene_id]
    transcript_ht = transcript_ht.select(
        transcript_expression=transcript_ht.row_value,
        expression_proportion=hl.struct(
            **{
                tissue: transcript_ht[tissue]/gene_prop[tissue]
                for tissue in tissues
            }
        )
    )
        
    return transcript_ht


def tissue_expression_struct_to_array(
    tissue_struct_expr: hl.expr.StringExpression,
    tissues: Optional[List[str]] = None,
    tissues_to_filter: Optional[List[str]] = None,
) -> Tuple[hl.expr.ArrayExpression, str]:
    if tissues is None:
        tissues = list(tissue_struct_expr.keys())

    if tissues_to_filter is not None:
        logger.info("Filtering tissues: %s", tissues_to_filter)
        tissues = [t for t in tissues if t not in tissues_to_filter]

    tissue_expression = hl.array([tissue_struct_expr[t] for t in tissues])

    return tissue_expression, tissues


def get_expression_proportion(
    tissue_struct_expr: hl.expr.StringExpression,
    tissues_to_filter: Optional[List[str]] = None,
) -> hl.expr.Float64Expression:
    tissue_expr, _ = tissue_expression_struct_to_array(
        tissue_struct_expr=tissue_struct_expr, 
        tissues_to_filter=tissues_to_filter
    )
    
    # Calculate the mean expression proportion across tissues.
    return hl.mean(hl.filter(lambda e: ~hl.is_nan(e), tissue_expr))

```

The other option is to change the summarize function back to only returning the transcript HT and then get the transcript proportion within this function instead of passing in 2 HTs. Something like this:

```
def summarize_transcript_expression(
    mt: hl.MatrixTable,
    transcript_expression_expr: Union[hl.expr.NumericExpression, str] = "transcript_tpm",
    tissue_expr: Union[hl.expr.StringExpression, str] = "tissue",
    summary_agg_func: Optional[Callable] = None,
) -> hl.Table:
    if summary_agg_func is None:
        summary_agg_func = lambda x: hl.median(hl.agg.collect(x))

    if isinstance(transcript_expression_expr, str):
        transcript_expression_expr = mt[transcript_expression_expr]

    if isinstance(tissue_expr, str):
        tissue_expr = mt[tissue_expr]

    mt = mt.group_cols_by(tissue=tissue_expr).aggregate(
        tx=summary_agg_func(transcript_expression_expr)
    )

    transcript_ht = mt.rename({"tx": ""}).make_table()
    transcript_ht = transcript_ht.key_by("transcript_id", "gene_id")
        
    return transcript_ht


def get_expression_proportion(
    transcript_ht: hl.Table,
    tissues_to_filter: Optional[List[str]] = None,
) -> hl.Table:
    transcript_ht = tissue_expression_ht_to_array(
        transcript_ht, tissues_to_filter=tissues_to_filter
    )
    
    tissues = hl.eval(transcript_ht.tissues)
    
    # Calculate the sum of transcript expression by gene per tissue.
    gene_ht = transcript_ht.group_by("gene_id").aggregate(
        expression_sum=hl.agg.array_sum(transcript_ht.tissue_expression)
    )

    # Calculate the mean expression proportion across tissues.
    transcript_ht = transcript_ht.annotate(
        exp_prop_mean=hl.mean(
            hl.filter(
                lambda e: ~hl.is_nan(e), 
                hl.map(
                    lambda x, y: x/y,
                    transcript_ht.tissue_expression,
                    gene_ht[transcript_ht.gene_id].expression_sum
                )
            )
        )
    )
    
    return transcript_ht

```

It sort of depends on whether it's worth storing the proportion as well (or if the HT with that info needs to be released), or if it can just be computed each time it is needed

Reply:
Im my original code, I put transcript_expression, gene_expression, and expression_proportion in one HT, it was clearer to me that way. And pext scores on v2 exomes are actually showing the expression_proportion, unless we will change that.